### ISSUE TITLE
astropy.wcs incompatible with wcslib 6.3

### ISSUE THREAD

**Issue by mboquien** [astropy/astropy#9085]
As mentioned by @privong in issue #9018, there is an error on import with astropy.wcs with wcslib version 6.3 and astropy 3.2.1. The backtrace is the following:

```
/usr/lib/python3.7/site-packages/astropy/wcs/__init__.py in <module>
     24 """
     25 
---> 26 from .wcs import *
     27 from . import utils
     28 

/usr/lib/python3.7/site-packages/astropy/wcs/wcs.py in <module>
     47 from astropy.io import fits
     48 from . import docstrings
---> 49 from . import _wcs
     50 
     51 from astropy.utils.compat import possible_filename

ImportError: /usr/lib/libwcs.so.6: undefined symbol: fits_read_wcstab
```

Thanks!

**Comment by nden** [comment]
@mboquien What os are you running this on? Did you build wcslib 6.3 or does it come with the system?

**Comment by jhunkeler** [comment]
(If this is Arch Linux) I think this might be a bug in the `wcslib` package. There should not be defining a `fits_read_wcstab` symbol in `libwcs.so` at all when `libcfitsio` is detected by the `configure` script

https://git.archlinux.org/svntogit/community.git/tree/trunk/PKGBUILD?h=packages/wcslib

I can't reproduce the missing symbol behavior when I build `cfitsio` and `wcslib` from source. However,
```
$ sudo pacman -S wcslib

$ nm -D /usr/lib/libwcs.so|grep fits_read_wcstab                                                                                                                                                                         
                 U fits_read_wcsta

# Just showing here it defined/provided in libcfitsio
$ nm -D /usr/lib/libcfitsio.so|grep fits_read_wcstab
00000000000e5d00 T fits_read_wcstab
```

I wonder if the `PKGBUILD` should have explicitly passed `--with-cfitsio[lib|inc]=` instead of letting `pkg-config` do it.

**Comment by jhunkeler** [comment]
https://github.com/astropy/astropy/blob/0e4db51fc8165147249daabf4513d0471d0249a2/cextern/wcslib/CHANGES#L50-L54

^ This probably has something to do with it too.

**Comment by privong** [comment]
@jhunkeler I'm the one who originally mentioned it, and yes it was on Arch Linux.

**Comment by jhunkeler** [comment]
@privong Ah ok, good.

Another odd thing... the `python-astropy` AUR package does not pass `--use-system-wcslib` yet somehow it ends up linking to it anyway, instead of doing the inline build.

For that to happen I assume the wcs extension must be passing `-lwcs` to the linker when it shouldn't be. As in, the linker sees it in `/usr/lib` first and uses it because its available.

I might have time to step through the build process tonight to see if thats what's happening.

**Comment by jhunkeler** [comment]
If you compile `wcslib` with FORTRAN support enabled the dangling symbol appears in `libwcs.so`. Their README isn't correct though -- `./configure --disable-fortran --with-cfitsiolib=/usr/lib --with-cfitsioinc=/usr/include --without-pgplot` produces a library without the undefined symbol. `--without-cfitsio` is was not enough to get rid of it as it suggest.

Anyway, Arch Linux's PKGBUILD does the following so let's just focus on that:
```bash
./configure --prefix=/usr --without-pgplot
```

This means it links with `cfitsio` by default (sort of). The library itself isn't linked to `cfitsio`:

```
$ ldd /usr/lib/libwcs.so
        linux-vdso.so.1 (0x00007ffca67f5000)
	libm.so.6 => /usr/lib/libm.so.6 (0x00007f6e72e0b000)
	libc.so.6 => /usr/lib/libc.so.6 (0x00007f6e72c48000)
	/usr/lib64/ld-linux-x86-64.so.2 (0x00007f6e730fa000)
```

It isn't useful to us but a program it provides does:

```bash
$ ldd /usr/bin/wcsware
... snip
	libcfitsio.so.8 => /usr/lib/libcfitsio.so.8 (0x00007fada44da000)
... snip

If you make a dummy C program at this point and try to link it with `-lwcs` it barfs:
```bash
$ gcc -I/usr/include -L/usr/lib -lwcs dummy.c                                                                                            
/usr/lib/libwcs.so: undefined reference to `fits_read_wcstab'
/usr/lib/libwcs.so: undefined reference to `gFitsFiles'
collect2: error: ld returned 1 exit status```

Passing `-Wl,--as-needed` to forces it succeed (Python's extension builder does this), but it doesn't change the fact the symbol is need is missing.

---

Building `wcslib` manually forcing it to link _everything_ with `libcfitsio` doesn't work either. The error morphed into:

```
$ gcc -I_install/tmp/wcslib/include -L_install/tmp/wcslib/lib -lwcs dummy.c
/usr/bin/ld: _install/tmp/wcslib/lib/libwcs.so: undefined reference to symbol 'fits_read_wcstab'
/usr/lib/libcfitsio.so.8: error adding symbols: DSO missing from command line
```

Linking the dummy program directly to `libwcs` and `libcfitsio` "just works":
```
$ gcc -lwcs -lcfitsio dummy.c
```

---

So what's Python doing? (i.e. `astropy/wcs/setup_package.py`)

```
$ python setup.py build --use-system-wcslib
...snip
gcc -pthread -shared -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now build/temp.linux-x86_64-3.7/astropy/wcs/src/distortion.o build/temp.linux-x86_64-3.7/astropy/wcs/src/distortion_wrap.o build/temp.linux-x86_64-3.7/astropy/wcs/src/docstrings.o build/temp.linux-x86_64-3.7/astropy/wcs/src/pipeline.o build/temp.linux-x86_64-3.7/astropy/wcs/src/pyutil.o build/temp.linux-x86_64-3.7/astropy/wcs/src/astropy_wcs.o build/temp.linux-x86_64-3.7/astropy/wcs/src/astropy_wcs_api.o build/temp.linux-x86_64-3.7/astropy/wcs/src/sip.o build/temp.linux-x86_64-3.7/astropy/wcs/src/sip_wrap.o build/temp.linux-x86_64-3.7/astropy/wcs/src/str_list_proxy.o build/temp.linux-x86_64-3.7/astropy/wcs/src/unit_list_proxy.o build/temp.linux-x86_64-3.7/astropy/wcs/src/util.o build/temp.linux-x86_64-3.7/astropy/wcs/src/wcslib_wrap.o build/temp.linux-x86_64-3.7/astropy/wcs/src/wcslib_tabprm_wrap.o -L/usr/lib -lwcs -lm -lpython3.7m -o build/lib.linux-x86_64-3.7/astropy/wcs/_wcs.cpython-37m-x86_64-linux-gnu.so
...snip
```

## tl;dr
```
-L/usr/lib -lwcs -lm -lpython3.7m 
```

The `_wcs` extension isn't linked to `cfitsio` so everything explodes. This might be a can of worms though... `cextern/cfitsio` isn't _exactly_ `cfitsio` and ensuring a fully viable build of `cfitsio` exists on the system is why the inline build exists in the first place.

Speaking of that, what does Python do without any `--use-system-*` arguments?

```
$ python setup.py build
...snip
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -march=x86-64 -mtune=generic -O3 -pipe -fno-plt -fPIC -DECHO -DWCSTRIG_MACRO -DASTROPY_WCS_BUILD -D_GNU_SOURCE -DNDEBUG -DHAVE_SINCOS -UDEBUG -I/home/jhunk/Downloads/apvenv/lib/python3.7/site-packages/numpy/core/include -Icextern/wcslib/C -Iastropy/wcs/include -I/home/jhunk/Downloads/apvenv/include -I/usr/include/python3.7m -c astropy/wcs/src/wcslib_tabprm_wrap.c -o build/temp.linux-x86_64-3.7/astropy/wcs/src/wcslib_tabprm_wrap.o -Wno-strict-prototypes -Wno-unused-function -Wno-unused-value -Wno-uninitialized
gcc -pthread -shared -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now build/temp.linux-x86_64-3.7/cextern/wcslib/C/flexed/wcsbth.o build/temp.linux-x86_64-3.7/cextern/wcslib/C/flexed/wcspih.o build/temp.linux-x86_64-3.7/cextern/wcslib/C/flexed/wcsulex.o build/temp.linux-x86_64-3.7/cextern/wcslib/C/flexed/wcsutrn.o build/temp.linux-x86_64-3.7/cextern/wcslib/C/cel.o build/temp.linux-x86_64-3.7/cextern/wcslib/C/dis.o build/temp.linux-x86_64-3.7/cextern/wcslib/C/lin.o build/temp.linux-x86_64-3.7/cextern/wcslib/C/log.o build/temp.linux-x86_64-3.7/cextern/wcslib/C/prj.o build/temp.linux-x86_64-3.7/cextern/wcslib/C/spc.o build/temp.linux-x86_64-3.7/cextern/wcslib/C/sph.o build/temp.linux-x86_64-3.7/cextern/wcslib/C/spx.o build/temp.linux-x86_64-3.7/cextern/wcslib/C/tab.o build/temp.linux-x86_64-3.7/cextern/wcslib/C/wcs.o build/temp.linux-x86_64-3.7/cextern/wcslib/C/wcserr.o build/temp.linux-x86_64-3.7/cextern/wcslib/C/wcsfix.o build/temp.linux-x86_64-3.7/cextern/wcslib/C/wcshdr.o build/temp.linux-x86_64-3.7/cextern/wcslib/C/wcsprintf.o build/temp.linux-x86_64-3.7/cextern/wcslib/C/wcsunits.o build/temp.linux-x86_64-3.7/cextern/wcslib/C/wcsutil.o build/temp.linux-x86_64-3.7/astropy/wcs/src/distortion.o build/temp.linux-x86_64-3.7/astropy/wcs/src/distortion_wrap.o build/temp.linux-x86_64-3.7/astropy/wcs/src/docstrings.o build/temp.linux-x86_64-3.7/astropy/wcs/src/pipeline.o build/temp.linux-x86_64-3.7/astropy/wcs/src/pyutil.o build/temp.linux-x86_64-3.7/astropy/wcs/src/astropy_wcs.o build/temp.linux-x86_64-3.7/astropy/wcs/src/astropy_wcs_api.o build/temp.linux-x86_64-3.7/astropy/wcs/src/sip.o build/temp.linux-x86_64-3.7/astropy/wcs/src/sip_wrap.o build/temp.linux-x86_64-3.7/astropy/wcs/src/str_list_proxy.o build/temp.linux-x86_64-3.7/astropy/wcs/src/unit_list_proxy.o build/temp.linux-x86_64-3.7/astropy/wcs/src/util.o build/temp.linux-x86_64-3.7/astropy/wcs/src/wcslib_wrap.o build/temp.linux-x86_64-3.7/astropy/wcs/src/wcslib_tabprm_wrap.o -L/usr/lib -lpython3.7m -o build/lib.linux-x86_64-3.7/astropy/wcs/_wcs.cpython-37m-x86_64-linux-gnu.so
...snip
```

This mode doesn't pass `-lwcs` to the linker, so it makes me wonder how Arch's `makepkg` managed to do it all by itself.

https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=python-astropy

**Comment by mboquien** [comment]
A few clarifications. The linked PKGBUILD corresponds to a revision I made to use the internal wcslib library rather than the system one in order to work around this issue. The PKGBUILD corresponding to this issue is https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=python-astropy&id=758a7265573c8ae83d05e6416cb87d56a99726a3 .

I looked at what links to what with the latest PKGBUILD,there is no link to wcslib (“find /usr/lib/python3.7/site-packages/astropy -name "*\.so" -exec ldd {} \; | grep wcs” returns nothing).

Should we ping the maintainer of wcslib?

**Comment by jhunkeler** [comment]
Thanks! I was completely baffled... I'm not sure if the wcslib maintainer needs a ping.

* `--use-system-cfitsio` links the FITS module against `libcfitsio`
* `--use-system-wcslib` links the WCS module against `libwcs`, and not `libcfitsio`.

In this case it feels like `astropy/wcs/setup_package.py` needs to be updated to link to `cfitsio`.

```python
# Might be nothing more than
cfg['libraries'] = ['cfitsio']
```

**Comment by nden** [comment]
@jhunkeler Won't this work only if the system `wcslib` was linked with `cfitsio` (which happens automatically if `cfitsio` is present but it may not be)? Is there a way to find out if this was the case?

**Comment by privong** [comment]
@nden The Arch package manager indicates that `wcslib` depends on `cfitsio`, so that dependency should be satisfied.

```
$ pacman -Qi wcslib
Name            : wcslib
Version         : 6.3-1
Description     : A C library that implements the 'World Coordinate System' (WCS) standard in FITS
[...]
Depends On      : cfitsio
[...]
```

**Comment by nden** [comment]
In an attempt to reproduce the problem on my RHEL, I built cfitsio. Then built wcslib with
```
export LDFLAGS="-L/path/to/cfitsio/lib -lcfitsio"
./configure --prefix=... --with-cfitsioinc=/path/to/cfitsio/include --with-cfitsiolib=/path/to/cfitsio/lib
```
Then built astropy with system libraries.
This was the only way to resolve the issue with undefined variables in astropy.wcs.

Would something like this work for the ARCH build? Note that I need to pass `-lcfitsio` to `LDFLAGS` when wcslib is built.
This is only necessary if cfitsio is available on the system and wcslib is built with it.

**Comment by nden** [comment]
FYI: I reported the problem to Mark Calabreta and he has just released wcslib v6.4 backing out the change that caused this issue.

**Comment by mboquien** [comment]
@nden Thanks. I have flagged the wcslib pacakge as out-of-date for Arch. As soon as it is updated, I will give it another try.

**Comment by nden** [comment]
@mboquien I'll keep the issue open. Please feel free to close if this is resolved.

**Comment by mboquien** [comment]
I tested with wcslib 6.4 and it appears to be working fine. Thanks!
